src/utils/ssr-utils.ts

```diff
@@ -26,84 +54,33 @@ export function reviveTaskInstance(instance: TaskInstance<any>) {
     instance._deferredObject.promise.finally(...params);
 }
 
-export function useTaskPrefetch<T>(
-  key: string,
-  task: Task<T, any>
-): TaskInstance<T> {
-  /* Server */
-  if (isServer()) {
-    // perform, add to prefetch, add to ssrContext
-    const taskInstance = task.perform();
-    onServerPrefetch(async () => {
-      try {
-        await taskInstance;
-        saveTaskToNuxtState(key, task);
-      } catch (e) {
-        // no need for extra handling
-      }
-    });
-    return taskInstance;
-  }
-
-  /* Client */
-  const [last] = reviveTaskInstances(key, task).reverse();
+function reviveTaskInstances (key, task) {
+  const taskCache = getNuxtTask(key);
 
-  if (last) {
-    return last;
-  } else {
-    return task.perform();
-  }
-}
-
-function saveTaskToNuxtState(key: string, task: Task<any, any>) {
-  const { $root } = getCurrentInstance() as any;
-  const nuxtState = $root && $root.context && $root.context.nuxtState;
-  if (!nuxtState) {
-    throw new Error("Could not access $root.context.nuxtState");
-  }
-
-  if (!nuxtState.vueConcurrency) {
-    nuxtState.vueConcurrency = {};
-  }
-
-  nuxtState.vueConcurrency[key] = computed(() => ({
-    instances: task._instances,
-  }));
-}
-
-function reviveTaskInstances(key: string, task: Task<any, any>) {
-  const taskCache = getTaskFromContext(key);
   if (taskCache) {
-    task._instances = taskCache.instances || [];
+    task._instances = taskCache || [];
     task._instances.forEach(reviveTaskInstance);
-    deleteTaskCache(key);
   }
-
-  return task._instances;
 }
 
-function getNuxtData() {
-  return (window as any).__NUXT__;
-}
+export function useTaskPrefetch (key, task) {
+  const vm = getCurrentInstance();
 
-function getTaskFromContext(key) {
-  if (!getNuxtData()) {
-    throw Error(`Could not access  window.__NUXT__`);
-  }
+  onServerPrefetch(async () => {
+    await wrap(task.perform);
+    setNuxtTask(vm, key, task);
+  });
 
-  return getNuxtData().vueConcurrency[key].value;
-}
+  if (process.client) {
+    onBeforeMount(async () => !task._instances.length && task.perform());
+    reviveTaskInstances(key, task);
```

Comment:
the revive happens after `task.perform()` ? shouldn't it be the other way around?

Reply:
In this case it's irrelevant because of `onBeforeMount`. In the alternative you proposed, you would need to revive the task instances prior to (conditionally) running `task.perform()`.